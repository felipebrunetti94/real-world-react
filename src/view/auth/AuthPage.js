// @ts-check

import PropTypes from "prop-types";
import Errors from "../error/Errors";

const AuthPage = ({
  title,
  getLink = () => <span>link</span>,
  showUsername = false,
  onSubmit,
  isLoading,
  error,
  updateAuthInfo,
  authInfo,
}) => {
  const handleChange = (e) => {
    const { name, value } = e.target;
    updateAuthInfo({ [name]: value });
  };

  const handleSubmit = (e) => {
    e.preventDefault();
    onSubmit(authInfo);
  };
  return (
    <div className="auth-page">
      <div className="container page">
        <div className="row">
          <div className="col-md-6 offset-md-3 col-xs-12">
            <h1 className="text-xs-center">{title}</h1>
            <p className="text-xs-center">{getLink()}</p>
            <Errors error={error} />
            <form onSubmit={handleSubmit}>
              {showUsername && (
                <fieldset className="form-group">
                  <input
                    onChange={handleChange}
                    value={authInfo.username || ""}
                    className="form-control form-control-lg"
                    type="text"
                    name="username"
                    placeholder="Your Name"
                    data-testid="input-username"
                  />
                </fieldset>
              )}
              <fieldset className="form-group">
                <input
                  onChange={handleChange}
                  value={authInfo.email || ""}
                  className="form-control form-control-lg"
                  type="text"
                  name="email"
                  placeholder="Email"
                  data-testid="input-email"
                />
              </fieldset>
              <fieldset className="form-group">
                <input
                  onChange={handleChange}
                  value={authInfo.password || ""}
                  className="form-control form-control-lg"
                  type="password"
                  name="password"
                  placeholder="Password"
                  data-testid="input-password"
                />
              </fieldset>
              <button
                className="btn btn-lg btn-primary pull-xs-right"
                type="submit"
                disabled={isLoading}
              >
                {title}
              </button>
            </form>
          </div>
        </div>
      </div>
    </div>
  );
};

AuthPage.propTypes = {
  title: PropTypes.string.isRequired,
  getLink: PropTypes.func,
  showUsername: PropTypes.bool,
  onSubmit: PropTypes.func,
  isLoading: PropTypes.bool,
  error: PropTypes.shape({
    message: PropTypes.string,
    details: PropTypes.arrayOf(PropTypes.string),
  }),
};

export default AuthPage;
