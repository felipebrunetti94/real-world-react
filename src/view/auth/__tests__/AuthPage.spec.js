import { fireEvent, getRoles, render, screen } from "@testing-library/react";
import AuthPage from "../AuthPage";

describe("View :: Auth :: AuthPage", () => {
  it("matches snapshot", () => {
    const props = {
      title: "test",
      showUsername: true,
      onSubmit: () => {},
      isLoading: false,
      errors: [],
      updateAuthInfo: () => {},
      authInfo: {},
    };
    const { container } = render(<AuthPage {...props} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  describe("when click on submit button", () => {
    it("call on submit with Auth Info", () => {
      const onSubmit = jest.fn();
      const authInfo = "auth info";
      const props = {
        title: "test",
        showUsername: true,
        onSubmit,
        isLoading: false,
        errors: [],
        updateAuthInfo: () => {},
        authInfo,
      };
      const { getByRole } = render(<AuthPage {...props} />);
      fireEvent.click(getByRole("button", { type: "submit" }));
      expect(onSubmit).toHaveBeenCalledWith(authInfo);
    });
  });

  describe("when update field", () => {
    it("call Update Auth Info with name and value of updated field", () => {
      const updateAuthInfo = jest.fn();
      const props = {
        title: "test",
        showUsername: true,
        onSubmit: () => {},
        isLoading: false,
        errors: [],
        updateAuthInfo,
        authInfo: {},
      };
      const { getByTestId } = render(<AuthPage {...props} />);

      fireEvent.change(getByTestId("input-username"), {
        target: {
          value: "text",
        },
      });

      expect(updateAuthInfo).toHaveBeenCalledWith({ username: "text" });
    });
  });

  describe("when receive Show User Name", () => {
    it("shows username field", () => {
      const props = {
        title: "test",
        showUsername: true,
        onSubmit: () => {},
        isLoading: false,
        errors: [],
        updateAuthInfo: () => {},
        authInfo: {},
      };
      const { getByTestId } = render(<AuthPage {...props} />);
      expect(getByTestId("input-username")).toBeInTheDocument();
    });
  });

  describe("when Is Loading", () => {
    it("submit button is disabled", () => {
      const props = {
        title: "test",
        showUsername: true,
        onSubmit: () => {},
        isLoading: true,
        errors: [],
        updateAuthInfo: () => {},
        authInfo: {},
      };
      const { getByRole } = render(<AuthPage {...props} />);

      expect(getByRole("button", { type: "submit" })).toBeDisabled();
    });
  });

  describe("when receive Errors", () => {
    it("show errors list", () => {
      const props = {
        title: "test",
        showUsername: true,
        onSubmit: () => {},
        isLoading: false,
        errors: [{ code: "1", message: "ohno!" }],
        updateAuthInfo: () => {},
        authInfo: {},
      };
      const { getByText } = render(<AuthPage {...props} />);

      expect(getByText("ohno!")).toBeInTheDocument();
    });
  });
});
