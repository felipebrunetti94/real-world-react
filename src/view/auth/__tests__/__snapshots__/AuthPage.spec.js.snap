// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`View :: Auth :: AuthPage matches snapshot 1`] = `
<div
  class="auth-page"
>
  <div
    class="container page"
  >
    <div
      class="row"
    >
      <div
        class="col-md-6 offset-md-3 col-xs-12"
      >
        <h1
          class="text-xs-center"
        >
          test
        </h1>
        <p
          class="text-xs-center"
        >
          <span>
            link
          </span>
        </p>
        <ul
          class="error-messages"
          data-testid="errors"
        />
        <form>
          <fieldset
            class="form-group"
          >
            <input
              class="form-control form-control-lg"
              data-testid="input-username"
              name="username"
              placeholder="Your Name"
              type="text"
              value=""
            />
          </fieldset>
          <fieldset
            class="form-group"
          >
            <input
              class="form-control form-control-lg"
              data-testid="input-email"
              name="email"
              placeholder="Email"
              type="text"
              value=""
            />
          </fieldset>
          <fieldset
            class="form-group"
          >
            <input
              class="form-control form-control-lg"
              data-testid="input-password"
              name="password"
              placeholder="Password"
              type="password"
              value=""
            />
          </fieldset>
          <button
            class="btn btn-lg btn-primary pull-xs-right"
            type="submit"
          >
            test
          </button>
        </form>
      </div>
    </div>
  </div>
</div>
`;
